src/imitation/util/logger.py

```diff
@@ -0,0 +1,197 @@
+        contexts are over.
+      format_strs: An list of output format strings that should be used by
+        every Logger initialized by this class during an `AccumulatingMeans`
+        context. For details on available output formats see
+        `stable_baselines.logger.make_output_format`.
+    """
+    self.default_logger = default_logger
+    self.current_logger = None
+    self._cached_loggers = {}
+    self._subdir = None
+    self.format_strs = format_strs
+    super().__init__(folder=self.default_logger.dir, output_formats=None)
+
+  @contextlib.contextmanager
+  def accumulate_means(self, subdir: str):
+    """Temporarily modifies this _HierarchicalLogger to accumulate means values.
+
+    During this context, `self.logkv(key, value)` writes the "raw" values in
+    "{self.default_logger.log_dir}/{subdir}" under the key "raw/{subdir}/{key}".
+    At the same time, any call to `self.logkv` will also accumulate mean values
+    on the default logger by calling
+    `self.default_logger.logkv_mean(f"mean/{subdir}/{key}", value)`.
+
+    During the context, `self.logkv(key, value)` will write the "raw" values in
+    `"{self.default_logger.log_dir}/subdir"` under the key "raw/{subdir}/key".
+
+    After the context exits, calling `self.dumpkvs()` will write the means
+    of all the "raw" values accumulated during this context to
+    `self.default_logger` under keys with the prefix `mean/{subdir}/`
+
+    Args:
+      subdir: A string key which determines the `folder` where raw data is
+        written and temporary logging prefixes for raw and mean data. Entering
+        an `accumulate_means` context in the future with the same `subdir`
+        will safely append to logs written in this folder rather than
+        overwrite.
+    """
+    if self.current_logger is not None:
+      raise RuntimeError("Nested `accumulate_means` context")
+
+    if subdir in self._cached_loggers:
+      logger = self._cached_loggers[subdir]
+    else:
+      folder = os.path.join(self.default_logger.dir, "raw", subdir)
+      os.makedirs(folder)
+      output_formats = _build_output_formats(folder, self.format_strs)
+      logger = sb_logger.Logger(folder, output_formats)
+      self._cached_loggers[subdir] = logger
+
+    try:
+      self.current_logger = logger
+      self._subdir = subdir
+      yield
+    finally:
+      self.current_logger = None
+      self._subdir = None
+
+  def logkv(self, key, val):
+    if self.current_logger is not None:
+      assert self._subdir is not None
+      raw_key = os.path.join("raw", self._subdir, key)
+      self.current_logger.logkv(raw_key, val)
+
+      mean_key = os.path.join("mean", self._subdir, key)
+      self.default_logger.logkv_mean(mean_key, val)
+    else:
+      self.default_logger.logkv_mean(key, val)
+
+  @property
+  def _logger(self):
+    if self.current_logger is not None:
+      return self.current_logger
+    else:
+      return self.default_logger
+
+  def logkv_mean(self, key, val):
```

Comment:
Will this nest correctly? e.g. say I was in a discriminator context (`"disc"`) and I did `.logkv_mean("foo", <value>)` `J` times, did a `.dumpkvs()`, then did `.logkv_mean()` `J` more times, and repeated that process `K` times. If I looked at the raw logs, I'd expect them to contain `K` entries of the form `raw/disc/foo: <mean of J values>`, while in the logs for the parent, I'd expect one entry of the form `mean/disc/foo: <mean of J*K values>`. As I understand it, the current implementation will write straight through to the "raw" child logger without putting anything in the parent, and without writing any key prefixes. Perhaps it's safer to remove this implementation & raise `NotImplementedError()` instead.

Reply:
Oh, on reflection the correct thing to do is probably raise `NotImplementedError` when `self.current_logger` is not `None`, else pass through to `self.default_logger`.

Reply:
I haven't thought too carefully about the inside-context `logkv_mean` case.

I'm a bit hesitant about raising `NotImplementedError` while inside the generator context. What if `gen_policy.learn()` (ie some arbitrary Stable Baselines algorithm) wants to call `logkv_mean` for some reason?.

Reply:
> As I understand it, the current implementation will write straight through to the "raw" child logger without putting anything in the parent, and without writing any key prefixes.

This understanding seems correct, thanks for catching this

Reply:
@qxcv 
How does this solution sound? Because we don't want to error out if some third-party package calls `log` or `logkv_mean` inside the context, and `HierachicalLogger` only claims to modify `logkv` behavior, we can just
(1) redirect `log`, `logkv_mean` to the default logger, unchanged and without prefixes and
(2) note this prominently in the documentation?